sw/device/lib/testing/usb_testutils_streams.c

```diff
@@ -0,0 +1,488 @@
+    s->rx_lfsr = rx_lfsr;
+  } else {
+    // In the event that we've received a zero-length data packet, we still
+    // must return the buffer to the pool
+    CHECK_DIF_OK(
+        dif_usbdev_buffer_return(usbdev->dev, usbdev->buffer_pool, &buf));
+  }
+}
+
+// Callback for successful buffer transmission
+static void strm_tx_done(void *stream_v, usb_testutils_xfr_result_t result) {
+  usbdev_stream_t *s = (usbdev_stream_t *)stream_v;
+  usb_testutils_streams_ctx_t *ctx = s->ctx;
+  usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+  // If we do not have at least one queued buffer then something has gone wrong
+  // and this callback is inappropriate
+  uint8_t tx_ep = s->tx_ep;
+  uint8_t nqueued = ctx->tx_bufs_queued[tx_ep];
+
+  if (s->verbose) {
+    LOG_INFO("strm_tx_done called. %u (%u total) buffers(s) are queued",
+             nqueued, ctx->tx_queued_total);
+  }
+
+  CHECK(nqueued > 0);
+
+  // Note: since buffer transmission and completion signalling both occur within
+  // the foreground code (polling, not interrupt-driven) there is no issue of
+  // potential races here
+
+  if (nqueued > 0) {
+    // Shuffle the buffer descriptions, without using memmove
+    for (unsigned idx = 1u; idx < nqueued; idx++) {
+      ctx->tx_bufs[tx_ep][idx - 1u] = ctx->tx_bufs[tx_ep][idx];
+    }
+
+    // Is there another buffer ready to be transmitted?
+    ctx->tx_queued_total--;
+    ctx->tx_bufs_queued[tx_ep] = --nqueued;
+
+    if (nqueued) {
+      CHECK_DIF_OK(
+          dif_usbdev_send(usbdev->dev, tx_ep, &ctx->tx_bufs[tx_ep][0u]));
+    }
+  }
+}
+
+// Callback for buffer reception
+static void strm_rx(void *stream_v, dif_usbdev_rx_packet_info_t packet_info,
+                    dif_usbdev_buffer_t buf) {
+  usbdev_stream_t *s = (usbdev_stream_t *)stream_v;
+  usb_testutils_streams_ctx_t *ctx = s->ctx;
+  usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+  CHECK(packet_info.endpoint == s->rx_ep);
+
+  // We do not expect to receive SETUP packets to this endpoint
+  CHECK(!packet_info.is_setup);
+
+  if (s->verbose) {
+    LOG_INFO("Stream %u: Received buffer of %u bytes(s)", s->id,
+             packet_info.length);
+  }
+
+  if (s->sending && s->generating) {
+    buffer_check(ctx, s, packet_info, buf);
+  } else {
+    // Note: this is useful for profiling the OUT performance on CW310
+    usb_testutils_ctx_t *usbdev = ctx->usbdev;
+
+    if (read_method != kReadMethodNone && packet_info.length > 0) {
+      alignas(uint32_t) uint8_t data[USBDEV_MAX_PACKET_SIZE];
+      size_t len = packet_info.length;
+      size_t bytes_read;
+
+      switch (read_method) {
+#if USBUTILS_MEM_FASTER
+        // Faster read performance using custom routine
+        case kReadMethodFaster: {
+          const uint32_t USBDEV_BUFFER_REG_OFFSET = 0x800U;
+          uint32_t offset =
+              USBDEV_BUFFER_REG_OFFSET + (buf.id * USBDEV_MAX_PACKET_SIZE);
+          usbutils_memcpy_from_mmio32(usbdev->dev->base_addr, offset, data,
```

Comment:
I could not find this function definition. I think you forgot to upload it, as this code is not compiled in CI.

Reply:
You can probably use `mmio_region_memcpy_from_mmio32` here.

Reply:
I haven't forgotten that function; it's in a separate PR on another branch, but it's inaccessible in this PR because the _MEM_FASTER macro is set to 0.

What that routine does is to introduce replacement routines for mmio_memcpy_from|to_mmio32 for the usbdev code because the generic function is very slow, but it also induces functional failures in DV chip sim so the replacement function is necessary. For now I switch it in manually, but really we need the replacement routines to be ratified and become the default.

The problems arise from the packet buffer memory being mapped as if it were CSRs, and it is indeed only word-addressable in the same fashion, but its behaviour differs because it must handle much more data in a performant manner, and there is no reset logic.